### PR TITLE
Add types for react-query

### PR DESCRIPTION
**Pull request by lukaszfiszer**
This PR includes typings for react-query package (npm: https://www.npmjs.com/package/react-query, github: https://github.com/tannerlinsley/react-query).

The package author does not plan to provide types to do the package - as discussed with https://github.com/tannerlinsley/react-query/issues/23.  

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If adding a new definition:
- [X] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [X] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [X] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [X] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [X] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [X] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/react-query/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+): QueryResult<TResult, TVariables>;
+
+export type QueryKey<TVariables> = string | [string, TVariables] | false | null | QueryKeyFunction<TVariables>;
+export type QueryKeyFunction<TVariables> = () => string | [string, TVariables] | false | null;
+
+export type QueryFunction<TResult, TVariables extends object> = (variables: TVariables) => Promise<TResult>;
+
+export interface QueryOptions<TResult> {
+    manual?: boolean;
+    pagination?: boolean;
```
Comment (tannerlinsley) [inline types/react-query/index.d.ts]: This should be paginated

File: types/react-query/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+export interface ReactQueryProviderConfig {
+    retry: number;
+    retryDelay: (attempt: number) => number;
+    staleTime: number;
+    cacheTime: number;
+    refetchAllOnWindowFocus: boolean;
+    refetchInterval: boolean;
+    suspense: boolean;
+}
```
Comment (tannerlinsley) [inline types/react-query/index.d.ts]: Add clearQueryCache method

**Review by tannerlinsley — CHANGES_REQUESTED** [review]

File: types/react-query/index.d.ts (hunk 1 above)
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: Fixed

File: types/react-query/index.d.ts (hunk 2 above)
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: Added

**Review by tannerlinsley — APPROVED** [review]

**Comment by PranavSenthilnathan** [comment]
Congratulations on your first DefinitelyTyped contribution! Thank you for being a part of the community!

**Pull request merged**